docs/design-docs/01-kpt-live-stdin.md

```diff
@@ -0,0 +1,242 @@
+- metadata.namespace: a valid namespace. default to value “default”
+- metadata.labels.cli-utils.sigs.k8s.io/inventory-id: A client-provided valid label.
+  Default value is a UUID e.g. 7c5af957-a3e2-4d68-8c0f-6c1864a66050
+
+### Convert Inventory to ResourceGroup
+
+The client-side Inventory is considered as the identifier of the cluster-side
+ResourceGroup CR. Right now, the Inventory is stored in the Kptfile to guarantee
+the ResourceGroup is unique. This requires Kptfile to exist to use `kpt live apply`.
+
+To split the Inventory from Kptfile to resourcegroup.yaml and convert the Inventory to
+ResourceGroup, `kpt live migrate` should be extended to map the inventory.name,
+inventory.namespace, inventory.inventoryID to  ResourceGroup CR metadata.name,
+metadata.namespace, metadata.labels.cli-utils.sigs.k8s.io/inventory-id correspondingly.
+
+Inventory in Kptfile
+```yaml
+apiVersion: kpt.dev/v1betaX
+kind: Kptfile
+inventory:
+  name: <INVENTORY_NAME>
+  namespace: <INVENTORY_NAMESPACE>
+  inventoryID: <INVENTORY_ID>
+```
+resoucegroup.yaml
+```yaml
+apiVersion: kpt.dev/v1alpha1
+kind: ResourceGroup
+metadata  
+  name: <INVENTORY_NAME>
+  namespace: <INVENTORY_NAMESPACE>
+  labels:
+    cli-utils.sigs.k8s.io/inventory-id: <INVENTORY_ID>
+```
+
+### ResourceGroup as a singleton object
+
+`kpt live apply [--resourcegroup-path] -` from STDIN accepts and only accepts a single
+ResourceGroup, including the ResourceGroup provided by the flag. It detects
+1. If more than one ResourceGroup is found, raise errors and display all the ResourceGroup objects.
+2. If no ResourceGroup is found in STDIN and and Kptfile inventory does not exists, raise errors and suggest users to
+   run kpt live init
+3. If no ResourceGroup is found in STDIN and Kptfile inventory exists, raise errors and
+   suggest users to run kpt live migrate
+
+## User Guide
+
+### To hydrate via kustomize and deploy via kpt
+
+#### Day 1
+
+<b>For new users to start from scratch (no Kptfile)</b>
+User can run `kpt live init [--resourcegroup-file=CUSTOM_RG.yaml]` to create a
+ResourceGroup object and store it in a resourcegroup.yaml file.
+Users can customize the file path with the flag “--resourcegroup-file”.
+
+<b>For existing kpt users to migrate from Kptfile</b>
+Users run `kpt live migrate [--resourcegroup-file=CUSTOM_RG.yaml]` to convert the
+Inventory object from Kptfile to a standalone resourcegroup.yaml file.
+Users can customize the file path with the flag “--resourcegroup-file”.
+
+<b>[optional]: Add shareable ResourceGroup to kustomize resources</b>
+If the ResourceGroup is expected to be shared in the Gitops workflow, users can add
+the resourcegroup.yaml  file path to the .resources field in kustomization.yaml.
+This simplifies the Day N deployment by omitting the “–resourcegroup-file“ flag.
+
+#### Day N
+
+- Users can configure the hydration rules in kustomization.yaml
+- Users can run `kustomize build <DIR> | kpt live apply -` to hydrate and deploy
```

Comment:
How does this work with `kpt fn source <DIR> | kpt live apply -`. Is `resourcegroup.yaml` treated as meta-resource or normal resource ?

Reply:
Thats a good point. Quick thought suggest - since info was part of Kptfile earlier that was meta-resource, should probably be considered as meta-resource, need to think more.

Reply:
qq: `kpt fn source` outputs a ResourceList CR, thus `kpt live apply -` won't deploy the k8s resources in ResourceList.items  but the actual ResourceList (and require the CRD to be preinstalled), right? Why do we suggest `kpt fn source <DIR> | kpt live apply -` here? @phanimarupaka is there any issues/discussions about it? 

We expect to have `kpt live apply` from STDIN be neutral to the hydration tool, so it does not assume the inputs are of any special kpt kind

Reply:
@yuwenma We need a way to pass inventory/RG information to `kpt live apply` through STDIN. But treating ResourceGroup as a meta resource will make functions modify it which is catastrophic. So we should treat ResourceGroup resource as a special resource and not modify it as part of functions and not include it in `kpt fn source`. In future, if there are use-cases which users come up with, we can include a new flag `--include-rg` for `kpt fn` commands. So for now, I am good with not supporting `kpt fn source <DIR> | kpt live apply -`.